app/src/visionglass/java/com/igalia/wolvic/FirstRunActivity.java

```diff
@@ -0,0 +1,71 @@
+package com.igalia.wolvic;
+
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.os.Bundle;
+
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentActivity;
+import androidx.fragment.app.FragmentTransaction;
+import androidx.preference.PreferenceManager;
+
+import com.igalia.wolvic.browser.SettingsStore;
+
+import java.util.Objects;
+
+public class FirstRunActivity extends FragmentActivity implements SharedPreferences.OnSharedPreferenceChangeListener {
+
+    private SharedPreferences mPrefs;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setContentView(R.layout.visionglass_first_run);
+        setTheme(R.style.Theme_MaterialComponents_Light);
+
+        mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
+        mPrefs.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        updateUI();
+    }
+
+    private void updateUI() {
+        if (!SettingsStore.getInstance(this).isTermsServiceAccepted()) {
+            Fragment fragment = BuildConfig.CN_FIRST_RUN_IN_PHONE_UI ? new FirstRunFragment() : new TermsServiceFragment();
+            getSupportFragmentManager().beginTransaction()
+                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
+                    .replace(R.id.fragment_placeholder, fragment)
+                    .commit();
+        } else if (!SettingsStore.getInstance(this).isPrivacyPolicyAccepted()) {
+            getSupportFragmentManager().beginTransaction()
+                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
+                    .replace(R.id.fragment_placeholder, new PrivacyPolicyFragment())
+                    .commit();
+        } else {
+            // finish and go to the main activity
+            Intent intent = new Intent(this, VRBrowserActivity.class);
+            startActivity(intent);
+            finish();
+        }
+    }
```

Comment:
What is not clear to me is what happens if the user declines to accept either the TOS or the privacy policy. If I am not mistaken, in that case no shared preference will change and thus this FirstActivity won't ever finish.

Reply:
We are reusing the functionality that is already implemented in `FirstRunFragment.onDisagreeButtonClicked()`.

If you decline the terms we show a second dialog, and if you decline that one then we finish the application. This is the same behaviour that we have in the HVR flavour.